### PR TITLE
Palmira.benavente/save reservation

### PR DESCRIPTION
**Pull request by PalmiraBenavente**
Save Reservation

- added funcionality for saving the reservation in a single class database.
- added mvp funcionality to dialogFragment Picker
- added class Reservation to join the information of reservation 
image
image
image

- video
https://user-images.githubusercontent.com/84336438/124799633-26a70580-df2b-11eb-8dc6-56cc3f571253.mp4

### CODE REVIEW

File: app/src/main/java/com/example/parkingsystem/listener/ListenerDialogFragmentDate.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.example.parkingsystem.listener;
+
+import java.util.Calendar;
+
+public interface ListenerDialogFragmentDate {
+    void setStartDateTime(Calendar calendarDateTime);
+    void setEndDateTime(Calendar calendarDateTime);
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem/listener/ListenerDialogFragmentDate.java]: both method I thing that do the same no?

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationParkingModel.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        if (reservation.getUserCode() == INT_MINUS_ONE) {
+            return false;
+        } else if (reservation.getParkingLot() == INT_MINUS_ONE) {
+            return false;
+        } else if (reservation.getStartDate() == null) {
+            return false;
+        } else if (reservation.getEndDate() == null) {
+            return false;
+        } else return true;
+    }
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem/mvp/model/ReservationParkingModel.java]: the validation off a reserve not be implemented in this PR , pls move it

**Comment by mauroLonghi** [comment]
Pls include a video to validate the right flow on the app.
How are you showing the date-time information? pls include any image

File: app/src/main/java/com/example/parkingsystem/listener/ListenerDialogFragmentDate.java (hunk 1 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/listener/ListenerDialogFragmentDate.java]: yes but no, each method follows the flow of each date, at the moment I can't think of how to unify them

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationParkingModel.java (hunk 2 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/model/ReservationParkingModel.java]: I removed this, done, the moment I did it, beyond the vadilation, in the next task, I thought that nulls could not be saved in the base independently of if they stepped or some other question

**Review by lucianoeli — APPROVED** [review]

File: app/src/main/java/com/example/parkingsystem/entity/Reservation.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private Calendar startDate;
+    private Calendar endDate;
+    private int parkingLot;
+    private int userCode;
+
+    public Reservation() {
+        this.parkingLot = INT_CERO;
+        this.userCode = INT_CERO;
+        this.startDate = null;
+        this.endDate = null;
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem/entity/Reservation.java]: pls, not assign a null in any field, instead off use an empty string

File: app/src/main/java/com/example/parkingsystem/mvp/view/DialogDateResarvationView.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public void setEndDateTime(ListenerDialogFragmentDate listenerDialogFragmentDate) {
+        listenerDialogFragmentDate.setEndDateTime(getCalendarDateTime());
+    }
+
+    private Calendar getCalendarDateTime() {
+        Calendar extractDateTime = new GregorianCalendar();
+        extractDateTime.set(binding.datePickerDialogFragment.getYear(),
+                binding.datePickerDialogFragment.getMonth() - 1,
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem/mvp/view/DialogDateResarvationView.java]: not use hardcodes pls

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java [hunk 5]
```diff
@@ -1,17 +1,24 @@
...
+import com.example.parkingsystem.R;
+import com.example.parkingsystem.databinding.ActivityReserverParkingLotBinding;
 import com.example.parkingsystem.fragments.DateReservationDialogFragment;
 import com.example.parkingsystem.mvp.contract.ReservationParkingContract;
 import com.example.parkingsystem.mvp.view.base.ActivityView;
 
 public class ReservationParkingView extends ActivityView implements ReservationParkingContract.ReservationActivityView {
 
     private final String DIALOG_FRAGMENT_DATE_PICKER = "dialog fragment date picker";
+    private final int INT_MINUS_ONE = -1;
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: I think that i saw this constant ... use a global static final int and replace in all the places pls

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java [hunk 6]
```diff
@@ -20,4 +27,45 @@ public void showDialogDataPiker() {
...
+        this.backToActivity();
+    }
+
+    @Override
+    public void toastShowTextDataError() {
+        Context context = getContext();
+        if (context != null)
+            Toast.makeText(context, context.getString(R.string.toast_resaervation_parking_error), Toast.LENGTH_SHORT).show();
+        this.backToActivity();
+    }
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: Could you use a generic method to unify the toast display?

Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: Done

File: app/src/main/java/com/example/parkingsystem/mvp/view/DialogDateResarvationView.java (hunk 4 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/view/DialogDateResarvationView.java]: Done

File: app/src/main/java/com/example/parkingsystem/entity/Reservation.java (hunk 3 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/entity/Reservation.java]: Done

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java (hunk 5 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: Done

File: app/src/main/java/com/example/parkingsystem/listener/ListenerDialogFragmentDate.java (hunk 1 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/listener/ListenerDialogFragmentDate.java]: Done

**Review by BQuinteros — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationParkingPresenter.java [hunk 7]
```diff
@@ -1,15 +1,53 @@
...
+
+    public void setReservation(Calendar calendarDateTime, boolean startDate) {
+        if (startDate) {
+            model.setStartDate(this.getStringDate(calendarDateTime));
+        } else model.setEndDate(this.getStringDate(calendarDateTime));
+    }
+
+    private String getStringDate(Calendar calendar) {
+        Date date = calendar.getTime();
+        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("hh: mm a dd-MMM-yyyy");
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationParkingPresenter.java]: do not hardcode, this string format should be into a constant

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java [hunk 8]
```diff
@@ -1,23 +1,63 @@
...
         FragmentManager fragmentManager = getFragmentManager();
-        DateReservationDialogFragment dateDialogFragment = DateReservationDialogFragment.newInstance();
+        DateReservationDialogFragment dateDialogFragment = DateReservationDialogFragment.newInstance(startDatePiker);
         dateDialogFragment.show(fragmentManager, DIALOG_FRAGMENT_DATE_PICKER);
     }
+
+    @Override
+    public int getParkingLot() {
+        String lot = binding.editTextReserverParkingPlace.getText().toString();
+        if (lot.isEmpty()) {
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: this logic shoudl be in the presenter, not here

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java [hunk 9]
```diff
@@ -1,23 +1,63 @@
...
+            return Constants.INT_MINUS_ONE;
+        } else
+            return Integer.parseInt(lot);
+    }
+
+    @Override
+    public int getParkingCode() {
+        String code = binding.editTextReserverParkingCode.getText().toString();
+        if (code.isEmpty()) {
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: same

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/DialogDateResarvationPresenter.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+import com.example.parkingsystem.mvp.contract.DialogDateResarvationContract;
+import com.example.parkingsystem.mvp.view.DialogDateResarvationView;
+
+public class DialogDateResarvationPresenter implements DialogDateResarvationContract.DialogFragmentDateReservationPresenter {
+    private DialogDateResarvationContract.DialogFragmentDateReservationView view;
+    private boolean startDatePicker;
+
+    public DialogDateResarvationPresenter(DialogDateResarvationView view, boolean startDatePicker) {
+        this.view = view;
+        this.startDatePicker = startDatePicker;
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/presenter/DialogDateResarvationPresenter.java:13]: why do you need this in the constructor?

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationParkingPresenter.java (hunk 7 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationParkingPresenter.java]: Done

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/DialogDateResarvationPresenter.java (hunk 10 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/presenter/DialogDateResarvationPresenter.java:13]: I associated it with the construction of the fragment, if I change this to the fragment it does not recognize what date it is

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java (hunk 8 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: Done

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java (hunk 9 above)
Comment (PalmiraBenavente) [reply app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: Done

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationParkingPresenter.java [hunk 11]
```diff
@@ -1,15 +1,61 @@
...
+        model.setReservationLotCode(this.getIntegerUserData(view.getParkingLot()), this.getIntegerUserData(view.getParkingCode()));
+        Reservation reservation = model.getReservation();
+        model.addReservation(reservation);
+        view.toastShowTextData(reservation.getParkingLot(), reservation.getUserCode(), reservation.getStartDate(), reservation.getEndDate());
+    }
+
+    public void setReservation(Calendar calendarDateTime, boolean startDate) {
+        if (startDate) {
+            model.setStartDate(this.getStringDate(calendarDateTime));
+        } else model.setEndDate(this.getStringDate(calendarDateTime));
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationParkingPresenter.java]: remember to use the curly brackets pls

File: app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java [hunk 12]
```diff
@@ -1,23 +1,55 @@
...
+    @Override
+    public String getParkingCode() {
+        return binding.editTextReserverParkingCode.getText().toString();
+    }
+
+    @Override
+    public void backToActivity() {
+        Activity activity = getActivity();
+        if (activity != null)
+            getActivity().finish();
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/view/ReservationParkingView.java]: here you have to use the activity variable that is the one that you know that is not null

**Review by SartiMau — CHANGES_REQUESTED** [review]

**Review by SartiMau — APPROVED** [review]

**Pull request merged**